src/full/Agda/TypeChecking/Irrelevance.hs

```diff
@@ -315,6 +315,54 @@ usableAtModality' ms why mod t =
 usableAtModality :: MonadConstraint TCM => WhyCheckModality -> Modality -> Term -> TCM ()
 usableAtModality = usableAtModality' Nothing
 
+{-# SPECIALIZE isNeverDefSing :: Type -> TCM DefSing #-}
+
+-- | Is the type definitely not a definitional singleton (even after
+--   substitution)? (Needs reduction.)
+isNeverDefSing :: (PureTCM m, MonadBlock m) => Type -> m DefSing
+isNeverDefSing a = do
+  let s = getSort a
+  p <- isProp <$> abortIfBlocked s
+  -- Strict propositions are always definitional singletons
+  if p
+  then pure MaybeDefSing
+  else do
+  -- I think the type should always be reduced by the time we get here...
+  -- a <- ignoreBlocking <$> reduceB a
+  case unEl a of
+    Var _ _           -> pure MaybeDefSing
+    Lam _ _           -> __IMPOSSIBLE__
+    Lit _             -> __IMPOSSIBLE__
+    Def d es          -> do
+      def <- theDef <$> getConstInfo d
+      case def of
+        Datatype{}         -> pure NeverDefSing
+        Record{recTel=tel} -> case recEtaEquality def of
+          YesEta  -> do
+            let Just vs = allApplyElims es
+            sings <- traverse (isNeverDefSing . unDom) (tel `apply` vs)
```

Comment:
This should filter out irrelevant fields, since an eta-record with only irrelevant fields is a definitional singleton.